libkbfs/folder_branch_ops.go

```diff
@@ -873,34 +909,46 @@ func (fbo *folderBranchOps) identifyOnce(
 	return nil
 }
 
-// if rtype == mdWrite || mdRekey, then mdWriterLock must be taken
-func (fbo *folderBranchOps) getMDLocked(
+// getMdLockedForRead returns an existing md for read
+// operation. Note that mds will not be fetched here.
+func (fbo *folderBranchOps) getMDLockedForRead(
 	ctx context.Context, lState *lockState, rtype mdReqType) (
 	md ImmutableRootMetadata, err error) {
+	if rtype != mdReadNeedIdentify && rtype != mdReadNoIdentify {
+		panic("Invalid rtype in getMDLockedForRead")
+	}
+
+	md = fbo.getTrustedHead(lState)
+	if md != (ImmutableRootMetadata{}) {
+		if rtype != mdReadNoIdentify {
+			err = fbo.identifyOnce(ctx, md.ReadOnly())
+		}
+		return md, err
+	}
+
+	return ImmutableRootMetadata{}, MDWriteNeededInRequest{}
+}
+
+// getMDLockedForWrite can fetch MDs, identify them and
+// contains the fancy logic. For reading use getMDLockedForRead.
+// Here we actually can fetch things from the server.
+// rekeys are untrusted.
+func (fbo *folderBranchOps) getMDLockedForWriteOrRekey(
+	ctx context.Context, lState *lockState, isRekey bool) (
+	md ImmutableRootMetadata, err error) {
 	defer func() {
-		if err != nil || rtype == mdReadNoIdentify || rtype == mdRekey {
+		if err != nil || isRekey {
 			return
 		}
 		err = fbo.identifyOnce(ctx, md.ReadOnly())
 	}()
 
-	md = fbo.getHead(lState)
+	md = fbo.getTrustedHead(lState)
 	if md != (ImmutableRootMetadata{}) {
 		return md, nil
 	}
 
-	// Unless we're in mdWrite or mdRekey mode, we can't safely fetch
-	// the new MD without causing races, so bail.
-	if rtype != mdWrite && rtype != mdRekey {
-		return ImmutableRootMetadata{}, MDWriteNeededInRequest{}
-	}
-
-	// We go down this code path either due to a rekey
-	// notification for an unseen TLF, or in some tests.
-	//
-	// TODO: Make tests not take this code path, and keep track of
-	// the fact that MDs coming from rekey notifications are
-	// untrusted.
+	// We come down this path in e.g. tests.
```

Comment:
is the old comment about coming down this code path due to a rekey notification for an unseen TLF not true now?